File: modules/angular2/src/mock/location_mock.js
import {SpyObject, proxy} from 'angular2/test_lib';

import {isBlank, isPresent, IMPLEMENTS} from 'angular2/src/facade/lang';
import {EventEmitter, ObservableWrapper} from 'angular2/src/facade/async';
import {List, ListWrapper} from 'angular2/src/facade/collection';
import {Location} from 'angular2/src/router/location';


@proxy
@IMPLEMENTS(Location)
export class DummyLocation extends SpyObject {
  urlChanges:List<string>;
  _path:string;
  _subject:EventEmitter;

  constructor() {
    super();
    this._path = '/';
    this.urlChanges = ListWrapper.create();
    this._subject = new EventEmitter();
  }

  setInitialPath(url:string) {
    this._path = url;
  }

  path():string {
    return this._path;
  }

  simulateUrlPop(pathname:string) {
    ObservableWrapper.callNext(this._subject, {
      'url': pathname
    });
  }

  go(url:string) {
    if (this._path === url) {
      return;
    }
    this._path = url;
    ListWrapper.push(this.urlChanges, url);
  }

  forward() {
    // TODO
  }

  back() {
    // TODO
  }

  subscribe(onNext, onThrow = null, onReturn = null) {
    ObservableWrapper.subscribe(this._subject, onNext, onThrow, onReturn);
  }

  noSuchMethod(m){return super.noSuchMethod(m);}
}

File: lib/src/mock/location_mock.dart
library angular2.src.mock.location_mock;

import "package:angular2/src/core/di.dart" show Injectable;
import "package:angular2/src/facade/async.dart"
    show EventEmitter, ObservableWrapper;
import "package:angular2/src/facade/collection.dart" show ListWrapper;
import "package:angular2/src/router/location.dart" show Location;

@Injectable()
class SpyLocation implements Location {
  List<String> urlChanges = [];
  /** @internal */
  String _path = "";
  /** @internal */
  String _query = "";
  /** @internal */
  EventEmitter<dynamic> _subject = new EventEmitter();
  /** @internal */
  String _baseHref = "";
  setInitialPath(String url) {
    this._path = url;
  }

  setBaseHref(String url) {
    this._baseHref = url;
  }

  String path() {
    return this._path;
  }

  simulateUrlPop(String pathname) {
    ObservableWrapper.callEmit(this._subject, {"url": pathname});
  }

  String prepareExternalUrl(String url) {
    if (url.length > 0 && !url.startsWith("/")) {
      url = "/" + url;
    }
    return this._baseHref + url;
  }

  go(String path, [String query = ""]) {
    path = this.prepareExternalUrl(path);
    if (this._path == path && this._query == query) {
      return;
    }
    this._path = path;
    this._query = query;
    var url = path + (query.length > 0 ? ("?" + query) : "");
    this.urlChanges.add(url);
  }

  forward() {}
  back() {}
  Object subscribe(dynamic /* (value: any) => void */ onNext,
      [dynamic /* (error: any) => void */ onThrow = null,
      dynamic /* () => void */ onReturn = null]) {
    return ObservableWrapper.subscribe(
        this._subject, onNext, onThrow, onReturn);
  }

  // TODO: remove these once Location is an interface, and can be implemented cleanly
  dynamic platformStrategy = null;
  String normalize(String url) {
    return null;
  }
}
